polaris/evaluate/_predictions.py

```diff
@@ -0,0 +1,197 @@
+
+
+class BenchmarkPredictions(BaseModel):
+    """
+    Base model to represent predictions in the Polaris code base.
+
+    Guided by [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle),
+    this class normalizes different formats to a single, internal representation.
+
+    Attributes:
+        predictions: The predictions for the benchmark.
+        target_labels: The target columns for the associated benchmark.
+        test_set_labels: The names of the test sets for the associated benchmark.
+    """
+
+    predictions: PredictionsType
+    target_labels: list[str]
+    test_set_labels: list[str]
+    test_set_sizes: dict[str, int]
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    @field_serializer("predictions")
+    def _serialize_predictions(self, predictions: PredictionsType):
+        """
+        Recursively converts all numpy values in the predictions dictionary to lists
+        so they can be serialized.
+        """
+
+        def convert_to_list(v):
+            if isinstance(v, np.ndarray):
+                return v.tolist()
+            elif isinstance(v, dict):
+                return {k: convert_to_list(v) for k, v in v.items()}
+
+        return convert_to_list(predictions)
+
+    @field_validator("target_labels", "test_set_labels")
+    @classmethod
+    def _validate_labels(cls, v: list[str]) -> list[str]:
+        if len(set(v)) != len(v):
+            raise ValueError("The predictions contain duplicate columns")
+        return v
+
+    @model_validator(mode="before")
+    @classmethod
+    def _validate_predictions(cls, data: dict) -> dict:
+        """Normalizes the predictions format to a standard representation we use internally"""
+
+        # This model validator runs before any Pydantic internal validation.
+        # This way we can normalize the incoming data to a standard representation.
+        # However, this implies that the fields can theoretically be any type.
+
+        # Ensure the type of the incoming predictions is correct
+        validator = TypeAdapter(IncomingPredictionsType, config={"arbitrary_types_allowed": True})
+        predictions = validator.validate_python(data.get("predictions"))
+
+        # Ensure the type of the target_labels and test_set_labels is correct
+        validator = TypeAdapter(list[str])
+        target_labels = validator.validate_python(data.get("target_labels"))
+        test_set_labels = validator.validate_python(data.get("test_set_labels"))
+
+        validator = TypeAdapter(dict[str, int])
+        test_set_sizes = validator.validate_python(data.get("test_set_sizes"))
+
+        # Normalize the predictions to a standard representation
+        predictions = cls._convert_lists_to_arrays(predictions)
+        predictions = cls._normalize_predictions(predictions, target_labels, test_set_labels)
+
+        # Normalize the predictions to a standard representation
+        cls._check_test_set_size(predictions, test_set_sizes)
```

Comment:
**Comment**
I feel like this would be a better for an **after** model validator. That would fit better with the intent of the Pydantic validator API.